packages/build-utils/src/fs/run-user-scripts.ts

```diff
@@ -687,69 +689,160 @@ export function getPathOverrideForPackageManager({
+    );
+  }
+
+  if (
+    validatedDetectedPackageManger.packageName !==
+    validatedCorepackPackageManager.packageName
+  ) {
+    throw new Error(
+      `Detected package manager "${validatedDetectedPackageManger.packageName}" does not match intended corepack defined package manager "${validatedCorepackPackageManager.packageName}". Change your lockfile or "package.json#packageManager" value to match.`
+    );
+  }
+
+  return intersects(
+    validatedDetectedPackageManger.packageVersionRange,
+    validatedCorepackPackageManager.packageVersionRange
+  );
+}
+
+function validateVersionSpecifier(version: string) {
+  if (!version) {
+    return undefined;
+  }
+
+  const [before, after, ...extra] = version.split('@');
+
+  if (extra.length) {
+    // should not have more than one `@`
+    return undefined;
+  }
+
+  if (!before) {
+    // should have a package before the `@`
+    return undefined;
+  }
+
+  if (!after) {
+    // should have a version after the `@`
+    return undefined;
+  }
+
+  if (!validRange(after)) {
+    // the version after the `@` should be a valid semver value
+    return undefined;
+  }
+
+  return {
+    packageName: before,
+    packageVersionRange: after,
   };
+}
 
+function detectPackageManager(
+  cliType: CliType,
+  lockfileVersion: number | undefined,
+  nodeVersion: NodeVersion | undefined
+) {
   switch (cliType) {
     case 'npm':
       switch (true) {
-        case corepackEnabled:
-          return no_override;
         case shouldUseNpm7(lockfileVersion, nodeVersion):
           return {
             path: '/node16/bin-npm7',
             detectedLockfile: 'package-lock.json',
-            detectedPackageManager: 'npm 7+',
+            detectedPackageManager: 'npm@7.x',
           };
         default:
-          return no_override;
+          return undefined;
       }
     case 'pnpm':
-      switch (detectPnpmVersion(lockfileVersion, corepackEnabled)) {
-        case 'corepack_enabled':
-          return no_override;
+      switch (detectPnpmVersion(lockfileVersion)) {
         case 'pnpm 7':
           // pnpm 7
           return {
             path: '/pnpm7/node_modules/.bin',
             detectedLockfile: 'pnpm-lock.yaml',
-            detectedPackageManager: 'pnpm 7',
+            detectedPackageManager: 'pnpm@7.x',
           };
         case 'pnpm 8':
           // pnpm 8
           return {
             path: '/pnpm8/node_modules/.bin',
             detectedLockfile: 'pnpm-lock.yaml',
-            detectedPackageManager: 'pnpm 8',
+            detectedPackageManager: 'pnpm@8.x',
           };
         case 'pnpm 9':
           // pnpm 9
           return {
             path: '/pnpm9/node_modules/.bin',
             detectedLockfile: 'pnpm-lock.yaml',
-            detectedPackageManager: 'pnpm 9',
+            detectedPackageManager: 'pnpm@9.x',
           };
         case 'pnpm 6':
         default:
-          return no_override;
+          return undefined;
```

Comment:
Question: Do we want to be more specific here with what was detected, but maybe leave the `path` blank because it matches the default? This would make the comparison between detected package managers and versions to corepack specified ones more complete.

Right now, detecting `pnpm@6.x` (for example) would come back as nothing, not comparing to what was specified by corepack.

Reply:
I'm in favor of that.

Reply:
[`2269bbc` (#11603)](https://github.com/vercel/vercel/pull/11603/commits/2269bbcf0698a0f9c875918c715925a24b526601)